### PR TITLE
Tweak syntax for page name refs

### PR DESCRIPTION
**Pull request by lubegasimon**
Fixes https://github.com/ocaml/odoc/issues/590

### CODE REVIEW

File: src/model/reference.ml [hunk 1]
```diff
@@ -128,6 +128,31 @@ let tokenize location s =
...
+           because of the scan_kind and scan_identifier calls that are attached to the parser.
+        *)
+        let f_occurence = String.index s '"' in
+        let l_occurence = String.rindex_from s f_occurence '"' in
+        let off_set = f_occurence + 1 in
+        let length = started_at - off_set - 1 in
+        let identifier = String.trim (String.sub s off_set length) in
+        (* The if ... then conditions specially handles cases like "foo"."bar",and
+           it should be checked first, otherwise, and exception will be thrown*)
+        if s.[0] = '"' && s.[f_occurence] = '"' then
```
Comment (lubegasimon) [inline src/model/reference.ml:149]: I’m still failing to handle a case like `{!”Foo”. “Bar”}` whose test is at https://github.com/ocaml/odoc/pull/622/commits/e4a86e69850fb694aa11fffadfae6030ecbd5a7f#diff-f59d447ba8018121335d4cfbebc893d052260724d49e9c187d9f4d7e2f7685ff

File: src/parser/token.ml [hunk 2]
```diff
@@ -66,7 +66,8 @@ type t =
   | `Begin_list_item of [ `Li | `Dash ]
   | `Minus
   | `Plus
-  | section_heading
+  | (* | `Quote *)
```
Comment (jonludlam) [inline src/parser/token.ml:69]: These should be removed

**Comment by jonludlam** [comment]
It'd be good to add some test cases to the parser tests - test/parser/test.ml, since when we remove the parser these tests will go with it.

File: src/model/reference.ml (hunk 1 above)
Comment (lubegasimon) [reply src/model/reference.ml:154]: Done!

File: src/parser/token.ml (hunk 2 above)
Comment (lubegasimon) [reply src/parser/token.ml]: Done!

File: src/model/reference.ml [hunk 3]
```diff
@@ -128,6 +128,41 @@ let tokenize location s =
...
+        let length = started_at - off_set - 1 in
+        let identifier = String.trim (String.sub s off_set length) in
+
+        (* The if ... then conditions specially handles cases like "foo"."bar",and
+            it should be checked first, otherwise, and exception will be thrown
+
+           However, I guarantee that this won't work expectedly for a case like
+           {!"Foo"."Bar".section-bar}
+        *)
+        if s.[0] = '"' && s.[f_occurence] = '"' then
```
Comment (Julow) [inline src/model/reference.ml:154]: This is wrong. (eg. `{!"Foo.1.2.3"."Bar.1.2.3"."Baz"}`).
The main mistake is `f_occurence`, which should be `index`. The parser works from right to left, `index` is the right quote and `l_occurence` is the corresponding left quote, nothing more needed.
The current value for `f_occurence` is the first quote of the string, we don't want that.

File: src/model/reference.ml [hunk 4]
```diff
@@ -128,6 +128,41 @@ let tokenize location s =
         scan_identifier started_at
           (open_parenthesis_count - 1)
           (index - 1) tokens
+    | '"' ->
```
Comment (Julow) [inline src/model/reference.ml:131]:
```suggestion
    | '"' when open_parenthesis_count = 0 ->
```

File: src/model/reference.ml [hunk 5]
```diff
@@ -128,6 +128,41 @@ let tokenize location s =
...
+           left to right. We do this specially to keep the identifier (page name in this case) intact.
+
+           However, the if ... else block operations, we return to then concept on how the parser works,
+           because of the scan_kind and scan_identifier calls that are attached to the parser.
+        *)
+        let f_occurence = String.index s '"' in
+        let l_occurence = String.rindex_from s f_occurence '"' in
+        let off_set = f_occurence + 1 in
+        let length = started_at - off_set - 1 in
+        let identifier = String.trim (String.sub s off_set length) in
```
Comment (Julow) [inline src/model/reference.ml:146]: `identifier_ended` is also checking that the identifier isn't `""`, `identifier_ended` should be called from here. (so you have the `location`, which will need adjustment to include the quotes)

File: src/model/reference.ml [hunk 6]
```diff
@@ -128,6 +128,41 @@ let tokenize location s =
...
+          let identifiers = String.split_on_char '.' s in
+          List.map
+            (fun ident ->
+              let ident =
+                String.trim (String.sub ident 1 (String.length ident - 2))
+              in
+              (None, ident, location))
+            identifiers
+        else
+          scan_kind identifier location (f_occurence - 1) (l_occurence - 1)
```
Comment (Julow) [inline src/model/reference.ml:164]: The `location` is the wrong one, see the comment about `identifier_ended` above.

We cannot call `scan_kind` without checking the next character first. Here we expect three things:
- `-`, to call `scan_kind`
- `.`, an identifier with no kind, recurse
- beginning of the string, an identifier with no kind, return

**Review by Julow — CHANGES_REQUESTED** [review]

**Comment by lubegasimon** [comment]
> It'd be good to add some test cases to the parser tests - test/parser/test.ml, since when we remove the parser these tests will go with it.

Indeed, I’ll surely address this in an independent PR

File: src/model/reference.ml [hunk 7]
```diff
@@ -128,11 +128,21 @@ let tokenize location s =
         scan_identifier started_at
           (open_parenthesis_count - 1)
           (index - 1) tokens
+    | '"' when open_parenthesis_count = 0 -> (
```
Comment (jonludlam) [inline src/model/reference.ml:131]: I don't believe we need the `when` clause here.

**Comment by jonludlam** [comment]
Thanks!

**Pull request merged**